flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.webmonitor.metrics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Nested data-structure to store metrics.
+ *
+ * This structure is not thread-safe.
+ */
+public class MetricStore {
+	private static final Logger LOG = LoggerFactory.getLogger(MetricStore.class);
+
+	JobManagerMetricStore jobManager = new JobManagerMetricStore();
+	Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
+	Map<String, JobMetricStore> jobs = new HashMap<>();
+
+	/**
+	 * Adds a metric to this MetricStore.
+	 *
+	 * @param name  the metric identifier
+	 * @param value the metric value
+	 */
+	public void add(String name, Object value) {
```

Comment:
You're right, I've combined two things here. But B) also depends a little bit on A) (at least implementation wise)

Concerning A) I'm not quite sure whether this is not a classical case of premature optimization. Have we actually measured the impact of the other approach? It would be good to know what it costs us. I would assume that the overhead is negligible.

Furthermore, I'm think you can also achieve an efficient serialization without too many short-living objects if you don't send strings around. First of all, we could serialize all metrics into a continuous byte buffer. Then we wouldn't have the object array overhead. Next we could use our own serializers to store the different object types efficiently (e.g. byte encoding for different metric types). On the server side you would only create the metric name objects once and that's all. Only upon deserialization you would have to create a new set of these objects. But that is ok since it happens on the web server side. The important thing is, though, to have a serialization component which is responsible for serialization and deserialization. That will avoid spreading the encoding over multiple places.